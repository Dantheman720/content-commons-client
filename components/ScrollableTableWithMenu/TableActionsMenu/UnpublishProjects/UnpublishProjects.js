import React from 'react';
import PropTypes from 'prop-types';
import { useMutation } from '@apollo/react-hooks';
import { Button, Popup } from 'semantic-ui-react';
import { UNPUBLISH_VIDEO_PROJECT_MUTATION } from 'lib/graphql/queries/video';
import { UNPUBLISH_PACKAGE_MUTATION } from 'lib/graphql/queries/package';

const UnpublishProjects = props => {
  const {
    team,
    handleResetSelections,
    handleActionResult,
    showConfirmationMsg,
    selections,
  } = props;

  // Determine which Query to run
  const setGraphQuery = () => {
    let query;
    const { contentTypes } = team;
    if ( contentTypes.includes( 'VIDEO' ) ) query = UNPUBLISH_VIDEO_PROJECT_MUTATION;
    if ( contentTypes.includes( 'PACKAGE' ) ) query = UNPUBLISH_PACKAGE_MUTATION;
    return query;
  };
  const graphQuery = setGraphQuery();

  const [unpublishDashboardProject, { loading, error }] = useMutation( graphQuery );

  const published = selections.filter( p => p.status === 'PUBLISHED' );

  const unpublishProject = async project => {
    const result = await unpublishDashboardProject( { variables: { id: project.id } } ).catch( error => ( {
      error,
      project,
      action: 'unpublish',
    } ) );
    handleActionResult( result );
  };

  const handleUnpublishProjects = async () => {
    await Promise.all( published.map( unpublishProject ) );
    handleResetSelections();
    showConfirmationMsg();
  };

  return (
    <Popup
      trigger={ (
        <Button
          className="unpublish"
          size="mini"
          basic
          onClick={ handleUnpublishProjects }
        >
          <span className="unpublish--text">Unpublish</span>
        </Button>
      ) }
      content="Unpublish Selection(s)"
      hideOnScroll
      inverted
      on={ ['hover', 'focus'] }
      size="mini"
    />
  );
};

UnpublishProjects.propTypes = {
  team: PropTypes.object,
  handleActionResult: PropTypes.func,
  handleResetSelections: PropTypes.func,
  showConfirmationMsg: PropTypes.func,
  selections: PropTypes.array,
};

export default UnpublishProjects;
