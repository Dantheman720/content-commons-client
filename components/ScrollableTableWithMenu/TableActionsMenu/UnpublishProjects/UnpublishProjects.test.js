import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Mutation } from 'react-apollo';
import { MockedProvider } from 'react-apollo/test-utils';
import { Button, Popup } from 'semantic-ui-react';
import UnpublishProjects from './UnpublishProjects';

const props = {
  handleResetSelections: jest.fn(),
  handleUnpublish: jest.fn(),
  handleUnpublishCacheUpdate: jest.fn(),
  showConfirmationMsg: jest.fn()
};

const Component = (
  <MockedProvider mocks={ [] } addTypename={ false }>
    <UnpublishProjects { ...props } />
  </MockedProvider>
);

describe( '<UnpublishProjects />', () => {
  it( 'renders without crashing', () => {
    const wrapper = mount( Component );
    const unpublishMutation = wrapper.find( UnpublishProjects );

    expect( unpublishMutation.exists() ).toEqual( true );
    expect( toJSON( unpublishMutation ) ).toMatchSnapshot();
  } );

  it( 'renders a Popup component', () => {
    const wrapper = mount( Component );
    const unpublishMutation = wrapper.find( UnpublishProjects );
    const popup = unpublishMutation.find( Popup );

    expect( popup.exists() ).toEqual( true );
    expect( popup.prop( 'content' ) ).toEqual( 'Unpublish Selection(s)' );
  } );

  it( 'renders a Button as its trigger', () => {
    const wrapper = mount( Component );
    const button = wrapper.find( Button );
    const btnTxt = <span className="unpublish--text">Unpublish</span>;

    expect( button.exists() ).toEqual( true );
    expect( button.prop( 'className' ) ).toEqual( 'unpublish' );
    expect( button.contains( btnTxt ) ).toEqual( true );
  } );

  it( 'clicking the Button calls handleUnpublish', () => {
    const wrapper = mount( Component );
    const button = wrapper.find( Button );

    button.simulate( 'click' );
    expect( props.handleUnpublish ).toHaveBeenCalled();
  } );

  it( 'mutation calls handleResetSelections and showConfirmationMsg on completion', async () => {
    const wrapper = mount( Component );
    const mutation = wrapper.find( Mutation );

    mutation.prop( 'onCompleted' )();
    expect( props.handleResetSelections ).toHaveBeenCalled();
    expect( props.showConfirmationMsg ).toHaveBeenCalled();
  } );

  it( 'mutation calls handleUnpublishCacheUpdate on update', async () => {
    const wrapper = mount( Component );
    const mutation = wrapper.find( Mutation );

    mutation.prop( 'update' )();
    expect( props.handleUnpublishCacheUpdate ).toHaveBeenCalled();
  } );
} );
