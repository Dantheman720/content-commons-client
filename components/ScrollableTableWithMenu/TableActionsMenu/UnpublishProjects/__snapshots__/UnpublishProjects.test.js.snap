// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnpublishProjects /> renders without crashing 1`] = `
<Apollo(UnpublishProjects)
  handleActionResult={[MockFunction]}
  handleResetSelections={[MockFunction]}
  selections={
    Array [
      Object {
        "author": Object {
          "firstName": "Edwin",
          "id": "cjrl1omfr002o075614zs8dxz",
          "lastName": "Mah",
        },
        "categories": Array [],
        "createdAt": "2019-06-17T15:23:46.184Z",
        "id": "cjx0j3ts9066y0708k1hg2g1b",
        "projectTitle": "Philippines Village",
        "status": "PUBLISHED",
        "team": Object {
          "id": "cjrkzhvku000f0756l44blw33",
          "name": "IIP Video Production",
          "organization": "Department of State",
        },
        "thumbnails": Array [
          Object {
            "alt": null,
            "id": "cjx0j3xqz067z0708frd7sd9g",
            "url": "2019/06/cjx0j3ts9066y0708k1hg2g1b/coverr-philippines-village-views-1559889358075.jpg",
          },
        ],
        "updatedAt": "2019-06-17T15:23:46.184Z",
        "visibility": "PUBLIC",
      },
      Object {
        "author": Object {
          "firstName": "Edwin",
          "id": "cjrl1omfr002o075614zs8dxz",
          "lastName": "Mah",
        },
        "categories": Array [],
        "createdAt": "2019-06-17T19:42:42.188Z",
        "id": "cjx0sctfx06ry0708bmtuf37f",
        "projectTitle": "Philippines Village 2",
        "status": "PUBLISHED",
        "team": Object {
          "id": "cjrkzhvku000f0756l44blw33",
          "name": "IIP Video Production",
          "organization": "Department of State",
        },
        "thumbnails": Array [
          Object {
            "alt": null,
            "id": "cjx0sd0yy06t30708by0b4suo",
            "url": "2019/06/cjx0sctfx06ry0708bmtuf37f/coverr-philippines-village-views-1559889358075.jpg",
          },
        ],
        "updatedAt": "2019-06-17T19:42:42.188Z",
        "visibility": "PUBLIC",
      },
    ]
  }
  showConfirmationMsg={[MockFunction]}
  unpublishVideoProject={[MockFunction]}
>
  <Mutation
    ignoreResults={true}
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "UNPUBLISH_VIDEO_PROJECT_MUTATION",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "unpublishVideoProject",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "ID",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 113,
          "start": 0,
        },
      }
    }
    variables={Object {}}
  >
    <UnpublishProjects
      handleActionResult={[MockFunction]}
      handleResetSelections={[MockFunction]}
      selections={
        Array [
          Object {
            "author": Object {
              "firstName": "Edwin",
              "id": "cjrl1omfr002o075614zs8dxz",
              "lastName": "Mah",
            },
            "categories": Array [],
            "createdAt": "2019-06-17T15:23:46.184Z",
            "id": "cjx0j3ts9066y0708k1hg2g1b",
            "projectTitle": "Philippines Village",
            "status": "PUBLISHED",
            "team": Object {
              "id": "cjrkzhvku000f0756l44blw33",
              "name": "IIP Video Production",
              "organization": "Department of State",
            },
            "thumbnails": Array [
              Object {
                "alt": null,
                "id": "cjx0j3xqz067z0708frd7sd9g",
                "url": "2019/06/cjx0j3ts9066y0708k1hg2g1b/coverr-philippines-village-views-1559889358075.jpg",
              },
            ],
            "updatedAt": "2019-06-17T15:23:46.184Z",
            "visibility": "PUBLIC",
          },
          Object {
            "author": Object {
              "firstName": "Edwin",
              "id": "cjrl1omfr002o075614zs8dxz",
              "lastName": "Mah",
            },
            "categories": Array [],
            "createdAt": "2019-06-17T19:42:42.188Z",
            "id": "cjx0sctfx06ry0708bmtuf37f",
            "projectTitle": "Philippines Village 2",
            "status": "PUBLISHED",
            "team": Object {
              "id": "cjrkzhvku000f0756l44blw33",
              "name": "IIP Video Production",
              "organization": "Department of State",
            },
            "thumbnails": Array [
              Object {
                "alt": null,
                "id": "cjx0sd0yy06t30708by0b4suo",
                "url": "2019/06/cjx0sctfx06ry0708bmtuf37f/coverr-philippines-village-views-1559889358075.jpg",
              },
            ],
            "updatedAt": "2019-06-17T19:42:42.188Z",
            "visibility": "PUBLIC",
          },
        ]
      }
      showConfirmationMsg={[MockFunction]}
      unpublishVideoProject={[Function]}
    >
      <Popup
        content="Unpublish Selection(s)"
        disabled={false}
        eventsEnabled={true}
        hideOnScroll={true}
        inverted={true}
        offset={0}
        on={
          Array [
            "hover",
            "focus",
          ]
        }
        pinned={false}
        position="top left"
        size="mini"
        trigger={
          <Button
            as="button"
            basic={true}
            className="unpublish"
            onClick={[Function]}
            size="mini"
          >
            <span
              className="unpublish--text"
            >
              Unpublish
            </span>
          </Button>
        }
      >
        <Portal
          closeOnDocumentClick={true}
          closeOnEscape={true}
          closeOnTriggerBlur={true}
          closeOnTriggerClick={false}
          closeOnTriggerMouseLeave={true}
          eventPool="default"
          mouseEnterDelay={50}
          mouseLeaveDelay={70}
          onClose={[Function]}
          onMount={[Function]}
          onOpen={[Function]}
          onUnmount={[Function]}
          openOnTriggerClick={false}
          openOnTriggerFocus={true}
          openOnTriggerMouseEnter={true}
          trigger={
            <Button
              as="button"
              basic={true}
              className="unpublish"
              onClick={[Function]}
              size="mini"
            >
              <span
                className="unpublish--text"
              >
                Unpublish
              </span>
            </Button>
          }
          triggerRef={
            Object {
              "current": <button
                class="ui mini basic button unpublish"
              >
                <span
                  class="unpublish--text"
                >
                  Unpublish
                </span>
              </button>,
            }
          }
        >
          <Ref
            innerRef={[Function]}
          >
            <RefFindNode
              innerRef={[Function]}
            >
              <Button
                as="button"
                basic={true}
                className="unpublish"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size="mini"
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <button
                        class="ui mini basic button unpublish"
                      >
                        <span
                          class="unpublish--text"
                        >
                          Unpublish
                        </span>
                      </button>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <button
                          class="ui mini basic button unpublish"
                        >
                          <span
                            class="unpublish--text"
                          >
                            Unpublish
                          </span>
                        </button>,
                      }
                    }
                  >
                    <button
                      className="ui mini basic button unpublish"
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="unpublish--text"
                      >
                        Unpublish
                      </span>
                    </button>
                  </RefFindNode>
                </Ref>
              </Button>
            </RefFindNode>
          </Ref>
        </Portal>
      </Popup>
    </UnpublishProjects>
  </Mutation>
</Apollo(UnpublishProjects)>
`;
